daffodil-io/src/main/scala/org/apache/daffodil/io/InputSource.scala

```diff
@@ -0,0 +1,472 @@
+   */
+  def areBytesAvailable(nBytes: Long): Boolean
+
+  /**
+   * Return the number of currently available bytes.
+   *
+   * This should not be used to determine the length of the data, as more bytes
+   * may become available in the future. This should really only be used for
+   * debug purposes.
+   */
+  def bytesAvailable(): Long
+
+  /**
+   * Return a single byte at the current byte position with a value in the
+   * range of 0 to 255.
+   *
+   * Increments the current byte position if successful.
+   *
+   * @return the byte at the current byte position if it exists, or -1 if EOF is reached.
+   */
+  def get(): Int
+
+  /**
+   * Return a byte array with data from the current byte position.
+   *
+   * Stores the next len bytes of data in dest starting at index off. In len
+   * bytes are not available or len bytes cannot fit in the dest array starting
+   * at the given offset, the dest array is not modified and false is returned.
+   *
+   * @return true if len bytes are available and written to the dest array,
+   *         false otherwise
+   */
+  def get(dest: Array[Byte], off: Int, len: Int): Boolean
+
+  /**
+   * Get the current byte position, using zero-based indexing
+   *
+   * @return the current byte position
+   */
+  def position(): Long
+
+  /**
+   * Set the current byte position, using zero-based indexing
+   *
+   * bytPos0b cannot be greater than the most recent read data. In other words,
+   * this can only be used to move backwards in data.
+   *
+   * @param bytePos0b the new current byte position
+   */
+  def position(bytePos0b: Long): Unit
+
+  /**
+   * Set the specified byte position as a location that that one may want to
+   * call setPosition in the future. This is essentially setting a mark in the
+   * data that can be reset back to later. Implementations are allowed to free
+   * any bytes before a locked byte position. It is implied tha the current
+   * byte position is locked, and thus any data that might exist after it
+   * cannot be freed.
+   *
+   * @param bytePos0b the byte position to lock
+   */
+  def lockPosition(bytePos0b: Long): Unit
+
+  /**
+   * Release a previously locked byte position, allowing the implementation to
+   * free any unlocked bytes.
+   *
+   * @param bytePos0b the byte position to release
+   */
+  def releasePosition(bytePos0b: Long): Unit
+
+  /**
+   * Alerts the implementation to attempt to free data that is no longer used,
+   * if possible. If possible, this should free any unlocked bytes.
+   */
+  def compact(): Unit
+
+  private var _debugging = false
+
+  final def areDebugging: Boolean = _debugging
+
+  final def setDebugging(setting: Boolean): Unit = _debugging = setting
+}
+
+/**
+ * Implements the InputSource interface, reading data from a
+ * generic java.io.InputStream and storing the data in buckets of a defined
+ * size. Buckets are freed when no locks exist inside the bucket to minimize
```

Comment:
Mixed terminology "lock" vs. "reference count".  Add comment explaining this has nothing to do with synchronization locks. The term "lock" here just means we've incremented the ref-count so that the block will not be released until ref count goes to zero.

Reply:
Fixed.